networkx/algorithms/centrality/betweenness.py

```diff
@@ -394,3 +397,78 @@ def _rescale_e(betweenness, n, normalized, directed=False, k=None):
         for v in betweenness:
             betweenness[v] *= scale
     return betweenness
+
+
+@not_implemented_for("graph")
+def _get_edge_keys(G, weight):
+    r"""Similar to `_weight_function`. Returns a function that
+    returns the keys of edges with the lowest weight between
+    two nodes.
+
+    This function is only suitable for multigraphs.
+
+    Parameters
+    ----------
+    G : NetworkX graph.
+
+    weight : string or function
+        If it is callable, `weight` itself is returned. If it is a string,
+        it is assumed to be the name of the edge attribute that represents
+        the weight of an edge. In that case, a function is returned that
+        gets the edge weight according to the specified edge attribute.
+
+    Returns
+    -------
+    function
+        This function returns a callable that accepts exactly three inputs:
+        a node `u`, a node adjacent to the first one `v` and the edge attribute
+        dictionary `d` for the edge joining those nodes. That function returns
+        a list with edge keys of those edges matching the lowest `weight` of
+        edges between `u` and `v`.
+
+    If any edge does not have an attribute with key `weight`, it is assumed to
+    have weight one.
+    """
+    _weight = _weight_function(G, weight)
+
+    def get_keys(u, v, d):
+        w = _weight(u, v, d)
+        return [k for k in d if d[k].get(weight, 1) == w]
+
+    return get_keys
+
+
+def _add_edge_keys(G, betweenness, weight=None):
+    r"""Adds the corrected betweenness centrality (BC) values for multigraphs.
+
+    Parameters
+    ----------
+    G : NetworkX graph.
+
+    betweenness : dictionary
+        Dictionary mapping adjacent node tuples to betweenness centrality values.
+
+    weight : string or function
+        See `_get_edge_keys` for details. Defaults to `None`.
+
+    Returns
+    -------
+    edges : dictionary
+        The parameter `betweenness` including edges with keys and their
+        betweenness centrality values.
+
+    The BC value is divided among edges of equal weight.
+    """
+    _keys = _get_edge_keys(G, weight)
+    betweenness.update(dict.fromkeys(G.edges(keys=True), 0.0))
+
+    for u, v in G.edges():
+        keys = _keys(u, v, G[u][v])
+        for k in keys:
+            betweenness[(u, v, k)] = betweenness[(u, v)] / len(keys)
+
+    for e in G.edges():  # Remove edges without key
+        if e in betweenness:
+            del betweenness[e]
```

Comment:
Nice! Thanks....
And  now it becomes more clear to me that we seem to be doing extra work here... We update the betweenness dict with 3-tuple edges and values 0 and then we update each of those. Finally we remove any dict entry with a 2-tuple edge.  BUT.... all the original entries will be deleted and all the new entries will be kept. So we could just make a new dict and avoid the extra adds and deletes.  Perhaps something like this?

```python
return {(u, v, k): betweenness[(u, v)] / len(keys) 
    for u, v in G.edges()
    for keys in [_keys(u, v, G[u][v])] 
    for k in keys
}
``` 
Notice the `for keys in ` loop uses brackets around the _keys to avoid having to call it twice inside a dict comprehension.

It might also be significantly faster to in-line the `_get_edge_keys` function since it is a private function and only used in this one loop.  Something like:
```python
    _weight = _weight_function(G, weight)
    result = {}
    for u, v in G.edges():
        d = G[u][v]
        wt = _weight(u, v, d)
        keys = [k for k in d if d[k].get(weight, 1) == wt]
        for k in keys:
            result[(u, v, k)] = betweenness[(u, v)] / len(keys)
    return result
```
If the timing isn't much different then use which ever is easier to read.

Reply:
Going from an empty dict would "forget" parallel edges with ```weight(e) > min(weights[(u,v)])``` . I also found that ```G.edges()``` will go over **all** edges but omit keys. If there are two edges ```(0,1,0), (0,1,1)``` the loop would go over ```(0,1)``` twice.

My solution now will create a new dict from the edges with keys and assign their BCs. I tried using a dict-comprehension because I'd favor its readability but it is vastly slower since we need to check for each key if we need to assign ```0.0``` or ```bc/len(keys)```. The variant submitted is the fastest alternative.

Reply:
Excellent!   Thanks!